### PR TITLE
feat: specified_id in dfx.json

### PR DESCRIPTION
**Pull request by lwshang**
# Description

In addition to passing `--specified-id` in `dfx deploy` and `dfx canister create`, `specified_id` can be set in `dfx.json`.

If it is set in both places, the specified ID from the command line takes precedence over the one in dfx.json.

Fixes SDK-1347

Also the duplicated logics of replacing `call_sender` with wallet is moved into `create_canister` function.

# How Has This Been Tested?

e2e tests: create.bash, deploy.bash

# Checklist:

- [ ] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] I have edited the CHANGELOG accordingly.
- [ ] I have made corresponding changes to the documentation.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,12 @@
 
 # UNRELEASED
 
+### feat: specified_id in dfx.json
+
+Aside from passing `--specified-id` in `dfx deploy` and `dfx canister create`, `specified_id` can be set in `dfx.json`.
```
Comment (ericswanson-dfinity) [inline CHANGELOG.md:7]:
```suggestion
In addition to passing `--specified-id` in `dfx deploy` and `dfx canister create`, `specified_id` can be set in `dfx.json`.
```

File: src/dfx/src/commands/canister/create.rs [hunk 2]
```diff
@@ -45,6 +43,7 @@ pub struct CanisterCreateOpts {
     ///
     /// This option only works with non-mainnet replica.
     /// This option implies the --no-wallet flag.
+    /// This option overwrites the specified_id field in dfx.json.
```
Comment (ericswanson-dfinity) [inline src/dfx/src/commands/canister/create.rs:46]:
```suggestion
    /// This option takes precedence over the specified_id field in dfx.json.
```

File: src/dfx/src/commands/deploy.rs [hunk 3]
```diff
@@ -68,6 +68,7 @@ pub struct DeployOpts {
     ///
     /// This option only works with non-mainnet replica.
     /// This option implies the --no-wallet flag.
+    /// This option overwrites the specified_id field in dfx.json.
```
Comment (ericswanson-dfinity) [inline src/dfx/src/commands/deploy.rs:71]: or "supercedes"

```suggestion
    /// This option takes precedence over the specified_id field in dfx.json.
```

File: src/dfx/src/lib/operations/canister/create_canister.rs [hunk 4]
```diff
@@ -85,6 +87,59 @@ pub async fn create_canister(
...
+                );
+            }
+            Some(specified_id_from_cli)
+        }
+        (Some(specified_id_from_json), None) => Some(specified_id_from_json),
+        (None, Some(specified_id_from_cli)) => Some(specified_id_from_cli),
+        (None, None) => None,
+    };
+
+    // Replace call_sender with wallet canister when necessary.
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/operations/canister/create_canister.rs:114]: Can you say more about the "when necessary" part?  
I'm not following why this is related to `specified_id.is_none()`

File: src/dfx/src/lib/operations/canister/create_canister.rs [hunk 5]
```diff
@@ -85,6 +87,59 @@ pub async fn create_canister(
...
+    let call_sender =
+        if specified_id.is_none() && !no_wallet && !matches!(call_sender, CallSender::Wallet(_)) {
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/operations/canister/create_canister.rs:116]: This expression has a lot of "not" in it.  I think we can rework it to be a little easier to follow, something like this (but please double-check the boolean logic):

```
if specified_id.is_some() || no_wallet || matches!(call_sender, Wallet...) {
    *call_sender
} else {
    match ...
}

```

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: src/dfx/src/lib/operations/canister/create_canister.rs [hunk 6]
```diff
@@ -85,6 +87,71 @@ pub async fn create_canister(
...
+    // If the network is IC mainnet, the specified ID will be overwritten by None.
+    if env.get_network_descriptor().is_ic && specified_id.is_some() {
+        warn!(
+            env.get_logger(),
+            "Specified ID is ignored on the IC mainnet."
+        );
+        specified_id = None;
+    }
+
+    // Replace call_sender with wallet canister unless:
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/operations/canister/create_canister.rs:123]: This comment describes the predicate below, and the flow, but I'm still not quite following.  Can you describe _why_ we need to replace the call sender with the wallet?

**Comment by lwshang** [comment]
As can be seen in the comment of `dfx canister --wallet` and `dfx deploy --wallet`:

> Specify a wallet canister id to perform the call.
If none specified, defaults to use the selected Identity's wallet canister.

That piece of code replacing `call_sender` was for this logic.

---

The `create_canister` method is now a complex shim which dispatch the creating canister logic to four underlying methods:
* reserve_canister_with_playground
* create_with_cycles_ledger
* create_with_management_canister
* create_with_wallet

There are many factors influence the dispatch:
* specify `--wallet`?
* set `--no-wallet`?
* which network: mainnet/playground/else?
* set `specified_id`?
* set `DFX_DISABLE_AUTO_WALLET`?
* `CYCLES_LEDGER_ENABLED`?
* maybe more...

The factors above are not orthogonal which makes it even more complicated.

We need a separate refactor PR to clear up and simplify the dispatch. I will file a Jira ticket.

**Pull request merged**